src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTokenExtensions.cs

```diff
@@ -170,6 +173,19 @@ public static bool IsBeginningOfStatementContext(this SyntaxToken token)
 
                 case SyntaxKind.ElseKeyword:
                     return true;
+
+                case SyntaxKind.CloseBracketToken:
+                    if (token.Parent.IsKind(SyntaxKind.AttributeList))
+                    {
+                        // attributes can belong to a statement
+                        var container = token.Parent.Parent;
+                        if (container is StatementSyntax)
```

Comment:
what about if you have this:

```
switch (c)
{
    case 0:
         [Foo]
         $$
}
```

The grandparent won't be a StatementSyntax there right?

Reply:
In this scenario the grandparent is an expression statement with a missing expression and missing semicolon.

Reply:
I think this is the case in general actually (for any "typing it in" scenario where you start with an attribute list at the end of a method).

Is there maybe a case where the *trailing* tokens could put us in a situation where the grandparent is not a statement?

Reply:
ah, gotha.  i thought the grandparent was the containing block.  that's why i wanted this switch case, in case it didn't work (since switch cases aren't statements).  If this is the expr-statement, i think this is ok.

> Is there maybe a case where the trailing tokens could put us in a situation where the grandparent is not a statement?

one general concern i have (and where we likely need tests) is what happens when we want keyword help here *and* there is code following that affects the parsing of `[Foo]`.  Can we add at least a few tests for common following constructs.  Likely something like:

```
[Foo]
$$
return bar;
```

```
[Foo]
$$
x y = bar();
```


```
[Foo]
$$
await bar;
```

```
[Foo]
$$
y = bar();
```

```
[Foo]
$$
bar();
```

This covers  a lot of cases where the parser can get confused based on what's coming before/after that point.

Reply:
It seems like the effect in practice is: we realize the grandparent is a statement (even if it's a return statement for example) and then when we decide whether we are in a local function context, we use the preceding, not trailing tokens.